pkg/util/metric: reduce allocations used when recording histogram metrics

abarganier opened the pull request:
Fixes: https://github.com/cockroachdb/cockroach/issues/116325

### Summary

When TSDB scrapes metrics registries, it doesn't store histograms in their entirety. Instead, we calculate a [predefined set of quantiles](https://github.com/cockroachdb/cockroach/blob/50a47aa0b457821da7ec3831f32eee8d397fe4d0/pkg/util/metric/metric.go#L1052-L1061) against the latest histogram window, and store the quantiles instead.

Previously, when calculating quantiles and averages, the metrics recorder and histogram library would allocate a new prometheus histogram for each calculation. This led to a massive spike in allocations caused by the metrics library twice every 10 seconds - once to record metrics into TSDB, and once to generate the node status. These allocations were frequently showing up in heap profiles, sometimes accounting for up to 30% of total objects allocation when the profile was timed properly. 

In addition to wasting memory, this also cause inconsistencies in our quantile, average, and sum/count calculations, since each was calculated against a potentially different view of the current histogram window. 

This PR optimizes the code that records histogram metrics to reduce allocations. It alters the interfaces used to always take in a histogram to calculate against as a parameter. This makes it obvious to the caller when they are allocation a prometheus histogram, and allows us to share a singly allocated histogram to be reused across calculations for consistency.

The PR also does some cleanup of the histogram interfaces, where certain functions have been made redundant now that we provide histograms as an argument to Mean/Total/Quantile calculations.

### Benchmark Results

Via benchmarking, it's revealed that these changes reduce the # of allocations for the metrics recorder's `extractValue()` function by **~84%**. 
```
// Master branch (without changes)
BenchmarkExtractValueAllocs
BenchmarkExtractValueAllocs-10         20692       56826 ns/op    107912 B/op     3526 allocs/op

// PR's branch (with changes)
BenchmarkExtractValueAllocs
BenchmarkExtractValueAllocs-10         87542       12169 ns/op     17288 B/op      566 allocs/op
```

This indicates that these changes accomplish their intended purpose, and that allocations caused by recording histogram metrics should be reduced significantly. 

### Follow Ups:

I've filed https://github.com/cockroachdb/cockroach/issues/116584 to address the quality of histograms within CRDB at a broader level. 

#### Note: please review this PR commit-wise for easier reading.

abarganier committed:
pkg/util/metric: don't rebuild histogram window for each quantile

Previously, when calculating quantiles for Histogram metrics, we would
build the windowed histogram separately for each quantile that we wanted
to calculate. We calculate 9 quantiles per-histogram, which means that
rebuilding the histogram windows for each quantile, for each histogram,
every 10 seconds, leads to a wasteful number of heap allocations.

Beyond being wasteful, it also means that each quantile calculated is
likely to be looking at a different histogram window, meaning that the
calculated quantiles are inconsistent from one another.

This patch updates the WindowedHistogram interface to include a function
to fetch the histogram window, and the ValueAtQuantileWindowed to take
the histogram window to calculate against as a param. This allows
callers to fetch the window *once* to calculate all quantiles against.

Functions like `Total()` and `MeanWindowed()` also need to benefit from
this work. A patch following this one will make those changes.

Release note: none

abarganier committed:
pkg/util/metric: make histogram Total() fn take histogram as argument

Previously, each time you called `Total()` on a windowed histogram, it
would allocate a new prometheus histogram. These allocations are
expensive. The API should force the caller to explicilty construct the
prometheus metric and pass them as arguments to functions that make
calculations against them, so that it's obvious to the user of the
histogram library when they're allocating.

This patch does so for the `Total()` function. It also adds TODOs to
some others (`Mean()`, `MeanWindowed()`, `TotalWindowed()`) to do
something similar. Those changes will come in following patches.

Release note: none

abarganier committed:
pkg/util/metric: make `Mean()` take in histogram, clean interface

Previously, `Mean()` and `MeanWindowed()` would allocate new prometheus
histograms on every call. This is wasteful from an allocations
perspective, and also makese the average calculated inconsistent with
any quantiles calculated after.

This patch updates `Mean()` to take in a histogram parameter, so the
caller is forced to explicitly allocate the prometheus histogram they
aim to calculate the average against.

The patch also collapses the `Mean()` and `MeanWindowed()` functions
together into the same function. The operations were exactly the same,
where the only difference was whether the cumulative or latest histogram
window was used. By rallying around `Mean()`, the same implementation
can be used for both, where the caller decides which based on which
histogram they provide as an argument.

Release note: none

abarganier committed:
pkg/util/metric: merge `TotalWindowed()` into `Total()`

The operations performed by `TotalWindowed()` and `Total()` are the
same. The only difference was whether the operation was performed
against the current histogram window, or the cumulative histogram.

Now that `Total()` takes the histogram to calculate against as a
parameter, `TotalWindowed()` is not necessary. Instead, we can just have
one function, and the caller can determine whether they want it
computed against the current window, or cumulative histogram,
by providing different arguments to `Total()`.

This helps reduce the surface area of the interface.

Release note: none

abarganier committed:
pkg/server/status: benchmark allocations used by recorder.extractValue()

Heap profiles have been showing that `extractValue()` is causing
unacceptably high numbers of allocations due to calculating
mean/sum/count/quantiles against histograms.

Future patches will clean up the `extractValue()` and histogram code to
reduce allocations when called. Once that work is done, this patch can
be used as a benchmark to quantify the reduction in allocations.

**Current Results:**
```
BenchmarkExtractValueAllocs
BenchmarkExtractValueAllocs-10         20692       56826 ns/op    107912 B/op     3526 allocs/op
```

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/116581)
<!-- Reviewable:end -->

abarganier commented:
### Manual Testing

Beyond unit testing & benchmarking, I tested things manually as well to ensure that reported quantiles still function correctly before & after these changes. Screenshots below show before/after results for each type of histogram (`Histogram`, `runtimeHistogram`, `ManualWindowHistogram`, `HdrHistogram`, and `AggHistogram`).

Before & after results show consistently reported quantiles, counts, sums, and averages across all histograms. 

#### `Histogram`
![normalhist](https://github.com/cockroachdb/cockroach/assets/8194877/dd0ec547-cf23-4347-8fe5-bbe051928fb3)

#### `runtimeHistogram`
![runtimehist](https://github.com/cockroachdb/cockroach/assets/8194877/a199a7a1-8aa5-4c6f-8f1a-e30e3c9ec66f)

#### `ManualWindowHistogram`
![manualhist](https://github.com/cockroachdb/cockroach/assets/8194877/21417fb5-d93b-4e2e-bf9f-67f2b35d7ead)

#### `AggHistogram`
##### Patched:
![agg_hist_patched](https://github.com/cockroachdb/cockroach/assets/8194877/4700d8b5-6edf-4eb9-83db-f78bd84124a6)

##### `master`:
![agg_hist_master](https://github.com/cockroachdb/cockroach/assets/8194877/0f9ca54f-e2c9-4043-aa14-80da53a86cc1)

petermattis reviewed (commented):
Nice reduction in allocations.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-:-NljOVD3AzyqUsQ4Znap:bwsb07f)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @dhartunian, @kvoli, and @miretskiy)

___
*[`pkg/util/metric/metric.go` line 433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljM__qD9fMh_yn157P:-NljM__qD9fMh_yn157Q:btswt70) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L433)):*
> ```Go
> 
> // ToPrometheusMetricWindowed returns a filled-in prometheus metric of the
> // right type for the current histogram window.
> ```

What is the difference between `ToPrometheusMetric()` and `ToPrometheusMetricWindowed()`? The comments and the naming don't make this clear. When should the caller use one vs the other.
___
*[`pkg/util/metric/metric.go` line 501 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljL2BwEFNiGhl5I5R7:-NljL2BwEFNiGhl5I5R8:b9p7lgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L501)):*
> ```Go
> //  2. Since the prometheus client library ensures buckets are in a strictly
> //     increasing order at creation, we do not sort them.
> func (h *Histogram) ValueAtQuantileWindowed(q float64, window *prometheusgo.Metric) float64 {
> ```

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).

I wonder if a better API would be for `ToPrometheusMetricWindowed()` to be renamed `Snapshot()` and return a `HistogramSnapshot struct {*prometheusgo.Histogram}` with `Total()`, `Mean()`, and `ValueAtQuantileWindowed()` methods.
___
*[`pkg/util/metric/metric.go` line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljLT0u9FGwFzlsZElh:-NljLT0u9FGwFzlsZEli:b-tt9ffy) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L677)):*
> ```Go
> // ToPrometheusMetricWindowedLocked returns a filled-in prometheus metric of the
> // right type.
> func (mwh *ManualWindowHistogram) ToPrometheusMetricWindowedLocked() *prometheusgo.Metric {
> ```

This method should not be exported as it isn't possible to use correctly outside of this package.
___
*[`pkg/util/metric/metric.go` line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljL_uG80OY_aF_tyoM:-NljL_uG80OY_aF_tyoN:b-kblyn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L718)):*
> ```Go
> // This function is very similar to Histogram.ValueAtQuantileWindowed. Thus see
> // Histogram.ValueAtQuantileWindowed for a more in-depth description.
> func (mwh *ManualWindowHistogram) ValueAtQuantileWindowed(
> ```

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).
___
*[`pkg/util/metric/metric.go` line 467 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljNdAw1q9iq5H2VJP0:-NljNdAw1q9iq5H2VJP1:b-kblyn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4067cd94433fc0fd8a5717663aa9f00a90c74389/pkg/util/metric/metric.go#L467)):*
> ```Go
> 
> // Total returns the (cumulative) number of samples and the sum of all samples.
> func (h *Histogram) Total(hist *prometheusgo.Metric) (int64, float64) {
> ```

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).
___
*[`pkg/util/metric/metric.go` line 473 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljNbeBDdWFWGTelLN4:-NljNbeBDdWFWGTelLN5:b-kblyn0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4067cd94433fc0fd8a5717663aa9f00a90c74389/pkg/util/metric/metric.go#L473)):*
> ```Go
> 
> // Mean returns the (cumulative) mean of samples.
> func (h *Histogram) Mean(hist *prometheusgo.Metric) float64 {
> ```

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 8 of 8 files at r1, 10 of 10 files at r2, 9 of 9 files at r3, 7 of 7 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116581)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @dhartunian, @kvoli, and @miretskiy)

___
*[`-- commits` line 78 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljPoOLECLESMZ89AFj):*
nit: If you reorder this commit as the first, it's easier to compare the before/after of your changes, for yourself and for anyone who wants to recreate your results.


<!-- Sent from Reviewable.io -->

abarganier commented:
*[`pkg/util/metric/metric.go` line 501 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljL2BwEFNiGhl5I5R7:-NljPplR2VktXqAEJErb:b-3gb6pd) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L501)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).

I wonder if a better API would be for `ToPrometheusMetricWindowed()` to be renamed `Snapshot()` and return a `HistogramSnapshot struct {*prometheusgo.Histogram}` with `Total()`, `Mean()`, and `ValueAtQuantileWindowed()` methods.
</blockquote></details>

This is a good idea, especially since the implementations for each of these functions has ended up being exactly the same for all implementations. I'll give it a try - it should play a nice role in resolving a lot of the other comments. Thanks!


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
I'm not convinced that this PR reduced allocations -- it looks to me that it perhaps moves allocation to the call site.

mgartner commented on pkg/ccl/changefeedccl/changefeed_processors.go:
```diff
@@ -1058,7 +1058,8 @@ func (j *jobState) checkpointCompleted(ctx context.Context, checkpointDuration t
 	j.lastProgressUpdate = j.ts.Now()
-	j.checkpointDuration = time.Duration(j.metrics.CheckpointHistNanos.Mean())
+	j.checkpointDuration = time.Duration(j.metrics.CheckpointHistNanos.Mean(
+		j.metrics.CheckpointHistNanos.ToPrometheusMetric()))
```
I think most of the savings are here: https://github.com/cockroachdb/cockroach/pull/116581/files#diff-8453981e212eb3eee4afbfde1f551dc981960870b369580d9b4f911b9108533cR704

We now call ToPrometheusMetricWindowed only once for the loop rather than `len(metric.RecordHistogramQuantiles)` times.

miretskiy replied:
ack.

abarganier replied:
Correct, key word I guess is the allocations are reduced when **recording** histogram metrics (as in, recording to TSDB). Agreed that the # of allocations in a use case like what we're doing here with the `CheckpointHistNanos` has the same number of allocations. 

I hear you re: ergonomics of the call sites, though. I have a patch coming that will help clean that up.

miretskiy replied:
Do we have to allocate every time we call ToPrometheusMetrics?

abarganier replied:
We do. We could use a `sync.Pool`, but the heavy allocations don't come from the initial construction of the objects. Instead, they come within 3rd party Prometheus code that "fills in" that allocated struct via a call to `Write()`: https://github.com/cockroachdb/cockroach/blob/master/pkg/util/metric/metric.go#L436

So, introducing a `sync.Pool` wouldn't have a significant impact on the problem we're trying to solve for this release blocker.

miretskiy replied:
I see... I thought this code changed things like recording a metric to require toPrometheusmetric call (that would be bad)... It didn't.  So, yeah, I agree -- sync.pool is not going to help.

abarganier committed:
pkg/util/metric: further clean WindowedHistogram interface

This patch aims to make the usage of windowed histograms vs. cumulative
histograms clearer & more straightforward. It decomposes much of the
`WindowedHistogram` interface, with a few notable changes:

- `HistogramSnapshot` is a new struct that's been added, which wraps a
  prometheus Histogram and isolates the implementations for `Mean()`,
`Total()`, and `ValueAtQuantile()` to a single place.
- `WindowedHistogram` is maintained, but only maintains a single
  function: `SnapshotWindowed()`. This returns a `HistogramSnapshot` for
  the current histogram window, from which further calculations can be
  made.
- `CumulativeHistogram` is a newly added interface, which sits alongside
  `WindowedHistogram`. It contains a single function,
  `CumulativeSnapshot`. This returns a `HistogramSnapshot` for the
  metric's cumulative histogram, from which further calculations can be
  made.

Combined, these changes aim to clarify which snapshot should be used
when, while removing the notion that things like mean, quantiles, sum,
and counts are calculated any differently between cumulative snapshots
and windowed snapshots (the operations are the same, only the dataset
changes). It also removes functions that calculated values against the
cumulative histogram from the `WindowedHistogram` interface, which was
confusing to say the least.

Latest benchmark values with these changes:
```
BenchmarkExtractValueAllocs
BenchmarkExtractValueAllocs-10    	  111093	      9459 ns/op	   17304 B/op	     566 allocs/op
```

This represents an ~84% improvement.

Release note: none

abarganier commented on pkg/util/metric/aggmetric/histogram.go:
```diff
@@ -184,7 +180,7 @@ type Histogram struct {
 
 // ToPrometheusMetric constructs a prometheus metric for this Histogram.
-func (g *Histogram) ToPrometheusMetric() *io_prometheus_client.Metric {
+func (g *Histogram) ToPrometheusMetric() *prometheusgo.Metric {
```
See above comment - I don't think using a `sync.Pool` would have a significant impact in this scenario.

abarganier reviewed (commented):
Thanks for the feedback all - RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-:-NlyYThr7-BMIakGpgzI:b8yw9m)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @dhartunian, @kvoli, @mgartner, @miretskiy, and @petermattis)

___
*[`-- commits` line 78 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljPoOLECLESMZ89AFj):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: If you reorder this commit as the first, it's easier to compare the before/after of your changes, for yourself and for anyone who wants to recreate your results.
</blockquote></details>

Good call! Done. 
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 1062 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116581#gh-1428565924:-NlyZfgs5YbACYP-rISm:boteqak) ([raw file](https://github.com/cockroachdb/cockroach/blob/4067cd94433fc0fd8a5717663aa9f00a90c74389/pkg/ccl/changefeedccl/changefeed_processors.go#L1062)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I see... I thought this code changed things like recording a metric to require toPrometheusmetric call (that would be bad)... It didn't.  So, yeah, I agree -- sync.pool is not going to help.
</blockquote></details>

:thumbsup: - appreciate the input. See the latest patch re: ergonomics improvements to the API - I hope it feels a bit less clunky!
___
*[`pkg/util/metric/metric.go` line 433 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljM__qD9fMh_yn157P:-NlyYZTdDPRUZp1UM4N8:bw5r406) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L433)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

What is the difference between `ToPrometheusMetric()` and `ToPrometheusMetricWindowed()`? The comments and the naming don't make this clear. When should the caller use one vs the other.
</blockquote></details>

With the `WindowedHistogram` and `CumulativeHistogram` interfaces, I think good progress has been made to disambiguate the two and make it obvious what each one represents, and via comments, which one should be used depending on the use case. LMK what you think.
___
*[`pkg/util/metric/metric.go` line 501 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljL2BwEFNiGhl5I5R7:-NlyYwNa895WHKJLt5sr:b-1pjz1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L501)):*
<details><summary><i>Previously, abarganier (Alex Barganier) wrote…</i></summary><blockquote>

This is a good idea, especially since the implementations for each of these functions has ended up being exactly the same for all implementations. I'll give it a try - it should play a nice role in resolving a lot of the other comments. Thanks!
</blockquote></details>

Thanks for this idea, I think it did a great job at cleaning things up a bit.

The latest patch is centered around the introduction of the `HistogramSnapshot` type, changes to the existing `WindowedHistogram` interface, and a new interface called `CumulativeHistogram`. Both interfaces now have the sole purpose of returning the appropriate `HistogramSnapshot`. LMK what you think. 
___
*[`pkg/util/metric/metric.go` line 677 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljLT0u9FGwFzlsZElh:-NlyZGX2DAbVJKHCf1iI:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L677)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This method should not be exported as it isn't possible to use correctly outside of this package.
</blockquote></details>

Fixed.
___
*[`pkg/util/metric/metric.go` line 718 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljL_uG80OY_aF_tyoM:-NlyZT8fDiNT7LIW1JEi:b-1gzif5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9406b0fe80f4793b54cef1fd496dd7a31a2c3192/pkg/util/metric/metric.go#L718)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).
</blockquote></details>

Pulled into `HistogramSnapshot`.
___
*[`pkg/util/metric/metric.go` line 467 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljNdAw1q9iq5H2VJP0:-NlyYsGm1mcmYUR071Qs:b-1gzif5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4067cd94433fc0fd8a5717663aa9f00a90c74389/pkg/util/metric/metric.go#L467)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).
</blockquote></details>

Pulled into `HistogramSnapshot`.
___
*[`pkg/util/metric/metric.go` line 473 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NljNbeBDdWFWGTelLN4:-NlyYvjn5vapGFvxGded:b-1gzif5) ([raw file](https://github.com/cockroachdb/cockroach/blob/4067cd94433fc0fd8a5717663aa9f00a90c74389/pkg/util/metric/metric.go#L473)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Is this method providing value anymore? It isn't using any fields of the receiver (`h`).
</blockquote></details>

Pulled into `HistogramSnapshot`.


<!-- Sent from Reviewable.io -->

petermattis reviewed (commented):
This looks a lot better, though there is an easy win to remove an unnecessary allocation from the new `HistogramSnapshot` code.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-:-NlyjtGa7OcoD2L6EhNA:b-xoye0j)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @abarganier, @dhartunian, @kvoli, @mgartner, and @miretskiy)

___
*[`pkg/util/metric/histogram_snapshot.go` line 32 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NlyiDqLAxOtT0UC796z:-NlyiDqLAxOtT0UC797-:b2tbpqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/5846a9aff0fd438b36fb757052b521a9cc095c73/pkg/util/metric/histogram_snapshot.go#L32)):*
> ```Go
> // NewHistogramSnapshot returns a new HistogramSnapshot instance, backed by the provided
> // Histogram.
> func NewHistogramSnapshot(h *prometheusgo.Histogram) *HistogramSnapshot {
> ```

You don't need to return a pointer here as we're never mutating `HistogramSnapshot.h` in any of the methods. Instead, the methods should be defined on `hs HistogramSnapshot` (i.e. a value receiver, not a pointer receiver). This removes an unnecessary allocation from this code path. Once this function is not returning a pointer it should be renamed to `MakeHistogramSnapshot` to follow our practice of using `New` to indicate allocation and `Make` to indicate initialization and returning of a struct by value.
___
*[`pkg/util/metric/histogram_snapshot.go` line 43 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-Nlyj1AhDP2FJrNwOy3G:-Nlyj1AhDP2FJrNwOy3H:b-o7480k) ([raw file](https://github.com/cockroachdb/cockroach/blob/5846a9aff0fd438b36fb757052b521a9cc095c73/pkg/util/metric/histogram_snapshot.go#L43)):*
> ```Go
> 	histogram := hs.h
> 	buckets := histogram.Bucket
> 	n := float64(*histogram.SampleCount)
> ```

Can `histogram.SampleCount` ever by `nil`? Should this be calling `histogram.GetSampleCount()` instead?
___
*[`pkg/util/metric/metric.go` line 99 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NlyjH1yD3LlAJ8Hujl2:-NlyjH1yD3LlAJ8Hujl3:bdfsy95) ([raw file](https://github.com/cockroachdb/cockroach/blob/4112c9a2c257dff61648d87d7c5601450970b7b8/pkg/util/metric/metric.go#L99)):*
> ```Go
> //
> // WindowedHistogram are generally only useful when recording histograms to TSDB,
> // where they are used the calculate quantiles and the mean. The exception is that
> ```

Nit: `s/used the/used to/g`


<!-- Sent from Reviewable.io -->

abarganier reviewed (commented):
Thanks for the feedback, very helpful. Ready for another look. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-:-Nlyukje6wr5cY74yGWa:b-tq790p)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @dhartunian, @kvoli, @mgartner, @miretskiy, and @petermattis)

___
*[`pkg/util/metric/histogram_snapshot.go` line 32 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NlyiDqLAxOtT0UC796z:-Nlyv7YX6x6XSereVrMc:bwtmln) ([raw file](https://github.com/cockroachdb/cockroach/blob/5846a9aff0fd438b36fb757052b521a9cc095c73/pkg/util/metric/histogram_snapshot.go#L32)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

You don't need to return a pointer here as we're never mutating `HistogramSnapshot.h` in any of the methods. Instead, the methods should be defined on `hs HistogramSnapshot` (i.e. a value receiver, not a pointer receiver). This removes an unnecessary allocation from this code path. Once this function is not returning a pointer it should be renamed to `MakeHistogramSnapshot` to follow our practice of using `New` to indicate allocation and `Make` to indicate initialization and returning of a struct by value.
</blockquote></details>

Done - Thanks for the tip (and cool seeing this in action with an updated benchmark result). Also appreciate the style guidelines/tips!
___
*[`pkg/util/metric/histogram_snapshot.go` line 43 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-Nlyj1AhDP2FJrNwOy3G:-NlyuymP0MfQItP47sWX:brq7j4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/5846a9aff0fd438b36fb757052b521a9cc095c73/pkg/util/metric/histogram_snapshot.go#L43)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can `histogram.SampleCount` ever by `nil`? Should this be calling `histogram.GetSampleCount()` instead?
</blockquote></details>

Ah, good eye - I missed this. Switched to `.GetSampleCount()`
___
*[`pkg/util/metric/metric.go` line 99 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NlyjH1yD3LlAJ8Hujl2:-NlyvJE9CtiBC3d0SI7q:b-x76dwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/4112c9a2c257dff61648d87d7c5601450970b7b8/pkg/util/metric/metric.go#L99)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Nit: `s/used the/used to/g`
</blockquote></details>

Oops - fixed.


<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-:-NlzP-8P8f_O1DfLnoTP:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi, @dhartunian, @kvoli, @mgartner, and @miretskiy)



<!-- Sent from Reviewable.io -->

abarganier commented:
TFTRs! 

bors r=petermattis

celiala commented:
https://github.com/cockroachdb/cockroach/issues/116325 was marked as a GA blocker, which means we need to also backport this to 23.2.0:

blathers backport 23.2.0-rc

abarganier commented:
@celiala the 23.2 backport for https://github.com/cockroachdb/cockroach/pull/114273 took so long to get approved that it didn't make it into the rc. I need to backport this first.

mgartner commented:
Nice clean up!





<!-- Sent from Reviewable.io -->

mgartner commented:
*[`pkg/util/metric/histogram_snapshot.go` line 21 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/116581#-NmD8cWg2Sjg3kVVgGBZ:-NmD8cWg2Sjg3kVVgGB_:b-dnwo2c) ([raw file](https://github.com/cockroachdb/cockroach/blob/276ad3d756fb54e5f27153770b6039a06e72be62/pkg/util/metric/histogram_snapshot.go#L21)):*
> ```Go
> 
> // HistogramSnapshot represents a point-in-time snapshot of a Histogram metric, against
> // which calculations like Mean, Total, and Quantiles can be calculated.
> ```

Nice job designing a clean, narrow API! I see this is data structure is tested in `metrics_test.go`, but you might consider adding unit tests specifically for this data structure in `histogram_snapshot.go`.


<!-- Sent from Reviewable.io -->
